ykrt/src/compile/jitc_yk/jit_ir.rs

```diff
@@ -70,11 +70,13 @@ impl<'a> LocalNumbers<'a> {
     /// # Panics
     ///
     /// Panics if the instruction isn't present in the module.
-    fn idx(&self, instr: &Instruction) -> Result<InstrIdx, CompilationError> {
+    fn idx(&self, instr: &Instruction) -> InstrIdx {
         // FIXME: This is inefficient.
         for (idx, candidate) in self.m.instrs().iter().enumerate() {
             if ptr::addr_eq(instr, candidate) {
-                return Ok(InstrIdx::new(idx)?);
+                // The `unwrap` cannot fail because we won't have expanded `m.instrs()` beyond the
+                // ability of `InstrIdx` to represent an index.
+                return InstrIdx::new(idx).unwrap();
```

Comment:
It's reasonable to assume we won't have more than 24^2 instructions, but I'm not sure we check that anywhere at the moment. In theory we could have a reaaaally long trace and then panic here. Ideally, we bail before that happens though.

Reply:
The JIT IR was explicitly designed so that index overflows could be temporary errors.

Reply:
Hang on. The comment actually says something interesting.

Is it true though? Does pushing an instruction beyond the index limit actually fail? I was under the impression it'd only fail if you try to materialise such an index. I don't think `TiVec` is aware of the fact that the index size is actually reduced (it relies on `usize` conversions, as I recall).

Reply:
With this diff we get to the panic:
```
diff --git a/ykrt/src/compile/jitc_yk/jit_ir.rs b/ykrt/src/compile/jitc_yk/jit_ir.rs
index f0f24fda..66ab9f5b 100644
--- a/ykrt/src/compile/jitc_yk/jit_ir.rs
+++ b/ykrt/src/compile/jitc_yk/jit_ir.rs
@@ -1279,7 +1279,7 @@ impl Module {
         let int8_type_idx = types.len().into();
         types.push(Type::Integer(IntegerType::new(8)));
 
-        Self {
+        let mut this = Self {
             name,
             instrs: Vec::new(),
             extra_args: Vec::new(),
@@ -1290,7 +1290,14 @@ impl Module {
             int8_type_idx,
             func_decls: TiVec::new(),
             global_decls: TiVec::new(),
+        };
+
+        for _ in 0..33554432 { // 2^25
+            this.push(LoadTraceInputInstruction::new(0, this.ptr_type_idx()).into());
         }
+        panic!("got here ok");
+
+        this
     }
 
     /// Returns the type index of [Type::Void].
```

Which shows that the instruction backing vector can happily grow larger than an `InstrIdx` can express when encoded into the JIT IR.

We can either make `push()` fail (but does anything ever shortcut `push()`?), or we can just always propagate an error when an instruction index is created. In theory it's OK for the vector to grow bigger than `InstrIdx` as long as you never need to reference one of the large indices.

I don't know what's best, but the comment is incorrect at present.

Reply:
> We can either make push() fail

This is what we should have done (and if we didn't, we should do).

Reply:
Go for it.

Reply:
(I reckon all of the other index types will have the same issue, so check all the `Module::push_*` methods also?)

Reply:
Fixed in 0e549469.